clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -708,11 +755,67 @@ protected Node curNode() {
             return curNode;
         }
 
+        public CallRetryContext callRetryContext() {
+            return callRetryContext;
+        }
+
+        private boolean shouldRetry(Throwable t) {
+            return t instanceof RetriableException || t instanceof UnsupportedVersionException;
+        }
+
+        /**
+         * Depending on what the exception is, we may choose to fail the Call, or retry it.
+         *
+         * @param failedCall    The failed call.
+         * @param t             The failure exception.
+         * @param now           The current time in milliseconds.
+         */
+        public void retryOrFail(Call failedCall, Throwable t, long now) {
+            if (shouldRetry(t)) {
+                failedCall.retry(failedCall, t);
+            } else {
+                failedCall.fail(now, t);
+            }
+        }
+
+        /**
+         * Handle a retry.
+         *
+         * Retry the call. This method takes in the failedCall which the retry call wants to
+         * base its CallRetryContext upon. If the number of retries exceeds the maximum allowed,
+         * the retry will fail with timeout. Sometimes the exception is thrown internally inside
+         * the client and we don't need a backoff.
+         *
+         * @param failedCall    The current time in milliseconds.
+         * @param throwable     The failure exception.
+         */
+        public void retry(Call failedCall, Throwable throwable) {
+            long now = time.milliseconds();
+            // If this is an UnsupportedVersionException that we can retry, do so. Note that a
+            // protocol downgrade will not count against the total number of retries we get for
+            // this RPC. That is why 'tries' is not incremented and retry backoff is not applicable.
+            if ((throwable instanceof UnsupportedVersionException) &&
+                    handleUnsupportedVersionException((UnsupportedVersionException) throwable)) {
+                log.debug("{} attempting protocol downgrade and then retry.", this);
+                runnable.call(this, now);
+                return;
+            }
+
+            CallRetryContext failedCallCallRetryContext = failedCall.callRetryContext();
```

Comment:
We can remove the extra call in the variable here.
```suggestion
            CallRetryContext failedCallRetryContext = failedCall.callRetryContext();
```

Reply:
Yeah, make sense. Refactored.